File: flask_app/app/services/discount_usage_service.py
from datetime import datetime
from app.dao.discount_usage_dao import DiscountUsageDAO

class DiscountUsageService:

    @staticmethod
    def get_all(page=1, per_page=10, filters=None):
        return DiscountUsageDAO.get_all(page=page, per_page=per_page, filters=filters or {})

    @staticmethod
    def get_by_id(usage_id):
        return DiscountUsageDAO.get_by_id(usage_id)

    @staticmethod
    def get_by_user(user_id, page=1, per_page=10):
        return DiscountUsageDAO.get_by_user(user_id, page, per_page)
    
    @staticmethod
    def get_user_usage_count(user_id):
        return DiscountUsageDAO.get_user_usage_count(user_id)
  
    @staticmethod
    def create( discount_id, user_id, applied_on, entity, discount_amount ):
        used_at = datetime.utcnow()
        data = { 
                "discount_id": discount_id, 
                "user_id": user_id, 
                "applied_on":applied_on,
                "entity": entity,
                "discount_amount": discount_amount, 
                "used_at": used_at 
            }
        
        return DiscountUsageDAO.create(data)

File: springboot_api/src/main/java/com/example/membership_api/service/DiscountUsageService.java
package com.example.membership_api.service;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.springframework.stereotype.Service;

import com.example.membership_api.constants.EntityType;
import com.example.membership_api.dao.DiscountUsageDAO;
import com.example.membership_api.model.DiscountUsage;

@Service
public class DiscountUsageService {

    private final DiscountUsageDAO discountUsageDAO;

    public DiscountUsageService(DiscountUsageDAO discountUsageDAO) {
        this.discountUsageDAO = discountUsageDAO;
    }

    public Map<String, Object> getAll(int page, int perPage, String userId, String discountId) {
        List<DiscountUsage> items = discountUsageDAO.getAll(page, perPage, userId, discountId);
        long total = discountUsageDAO.getTotal(userId, discountId);

        Map<String, Object> response = new HashMap<>();
        response.put("items", items);
        response.put("total", total);
        response.put("page", page);
        response.put("perPage", perPage);
        return response;
    }

    public DiscountUsage getById(String id) {
        return discountUsageDAO.getById(id);
    }

    public List<DiscountUsage> getByUser(int page, int perPage, String userId) {
        return discountUsageDAO.getByUser(userId, page, perPage);
    }

    public long getUserUsageCount(String id) {
        return discountUsageDAO.getUserUsageCount(id);
    }


    public DiscountUsage create(String discountId, String userId, String appliedOn, EntityType entity, Double discountAmount) {
        DiscountUsage usage = new DiscountUsage();
        usage.setDiscountId(discountId);
        usage.setUserId(userId);
        usage.setAppliedOn(appliedOn);
        usage.setEntity(entity);
        usage.setDiscountAmount(discountAmount);
        usage.setUsedAt(LocalDateTime.now());
        return discountUsageDAO.create(usage);
    }
}
